clang/lib/Sema/SemaChecking.cpp

```diff
@@ -1900,13 +1934,157 @@ Sema::CheckBuiltinFunctionCall(FunctionDecl *FDecl, unsigned BuiltinID,
     if (Context.getTargetInfo().areAllPointersCapabilities()) {
       Diag(TheCall->getBeginLoc(), diag::err_builtin_target_unsupported)
           << TheCall->getSourceRange();
       return ExprError();
     }
+    // First argument is the authorizing capability, the second is the
+    // capability to be converted to a relative integer pointer.
+    // For this builtin we reject __(u)intcap_t arguments as the second arg.
+    QualType SrcTy;
+    if (checkArgCount(*this, TheCall, 2) || checkCapArg(*this, TheCall, 0) ||
+        checkCapArg(*this, TheCall, 1, &SrcTy))
+      return ExprError();
+    if (!SrcTy->isPointerType())
+      return ExprError(Diag(TheCall->getBeginLoc(),
+                            diag::err_typecheck_expect_pointer_operand)
+                       << SrcTy << TheCall->getSourceRange());
+    TheCall->setType(Context.getPointerType(SrcTy->getPointeeType(),
+                                            ASTContext::PIK_Integer));
+    break;
+  }
+  case Builtin::BI__builtin_cheri_cap_from_pointer: {
+    // First argument is the authorizing capability, the second is the
+    // integer to be converted to a capability.
+    // For purecap the second argument must be a (non-capability) integer type
+    // and in hybrid mode we also permit (non-capability) pointer types.
+    // If the non-capability argument is a pointer the return type is the
+    // will be the capability equivalent of that pointer, if it's an integer
+    // the return type is void* __capability.
+    if (checkArgCount(*this, TheCall, 2) || checkCapArg(*this, TheCall, 0))
+      return ExprError();
+    auto PtrArg = TheCall->getArg(1);
+    auto PtrTy = PtrArg->getType();
+    QualType ResultPointee = Context.VoidTy;
+    if (!PtrTy->isIntegerType() || PtrTy->isIntCapType()) {
+      if (Context.getTargetInfo().areAllPointersCapabilities()) {
+        Diag(PtrArg->getExprLoc(), diag::err_typecheck_expect_int) << PtrTy;
+        return ExprError();
+      } else if (PtrTy->isPointerType() &&
+                 !PtrTy->isCHERICapabilityType(Context)) {
+        ResultPointee = PtrTy->getPointeeType();
+      } else {
+        Diag(PtrArg->getExprLoc(), diag::err_typecheck_expect_scalar_operand)
+            << PtrTy;
+        return ExprError();
+      }
+    }
+    TheCall->setType(
+        Context.getPointerType(ResultPointee, ASTContext::PIK_Capability));
+    break;
+  }
+  case Builtin::BI__builtin_cheri_perms_check:
+    // Overloaded to allow any capability type as the first argument.
+    if (checkArgCount(*this, TheCall, 2) || checkCapArg(*this, TheCall, 0) ||
+        checkBuiltinArgument(*this, TheCall, 1))
+      return ExprError();
     break;
+  case Builtin::BI__builtin_cheri_type_check:
+    // Overloaded to allow any capability type as the first and second argument.
+    if (checkArgCount(*this, TheCall, 2) || checkCapArg(*this, TheCall, 0) ||
+        checkCapArg(*this, TheCall, 1))
+      return ExprError();
+    break;
```

Comment:
This is the same code as subset test? One's an assertion, the other's a query, but maybe they should share code here?

Reply:
Yeah that probably makes sense. I just placed it below the other assert because I made the changes separate and then cherry-picked.